rfcs/0004-transaction.md

```diff
@@ -0,0 +1,291 @@
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Isolation Levels
+
+The Snapshot Isolation have a good reputation in the industry, and is widely used in many databases. It's gives a lightweight Snapshot on each transactions started, and checks whether the writes against the snapshot has modified by others or not during the commit. It is very efficient and is considered as a good default isolation level for most of the use cases.
+
+However, Snapshot Isolation still suffers Write Skew anomaly. For example, let's assume we have two accounts, where account A has $600, and account B has $500. Also, the business rule is that Account A's balance + Account B's balance should always be >= $200.
+
+Two transactions occur concurrently:
+
+1. Transaction 1 first checks the total amount of balance in A and B, and then transfers $550 from account A to account C.
+2. Transaction 2 first checks the total amount of balance in A and B, and then transfers $450 from account B to account D.
+
+In this case, if the two transactions are executed concurrently, the total amount of balance in A and B might be less than $200, which violates the business rule. There's no Write-Write conflict in the two transactions, but the Read-Write conflict is not detected by Snapshot Isolation.
+
+To mitigate this anomaly, we could choose:
+
+1. use Serializable Snapshot Isolation (SSI) level: increase the isolation level, and no longer allow the write skew anomaly to occur.
+2. use `get_for_update()` to synchronize the critical read operations and write operations in the SI transaction.
+
+(1) is the approach that Badger takes, it supported a lightweight SSI transaction implementation.
+
+(2) is the approach that RocksDB takes, it does not built in SSI support, but it provides a `GetForUpdate()` method to lock the key before the write operation.
+
+In my understanding, the pros of the RocksDB approach is mostly easy to implement and efficient, and free of floating garbage (when compares with innodb's UNDO logs approach and the Postgres's multi version rows approach). Also, in most of the time, users can make the decision by themselves about when to use `GetForUpdate()` to make the critical concurrency control correct.
```

Comment:
Don't we need multi-version rows to support snapshot isolation? E.g. if a transaction is pegged to some snapshot at sequence number N, we need to always return keys with sequence number <= N even if those keys are updated. Or are you just saying that this is only required for the memtable with the rocksdb approach?

Reply:
yes indeed, multi-version rows with sequence number must be a prerequisites on this.

i'll move the description about rocksdb' approach into an "other systems" section in this rfc doc as @vigneshc suggested.